aws/rust-runtime/aws-runtime/src/invocation_id.rs

```diff
@@ -9,58 +9,132 @@ use aws_smithy_runtime_api::client::interceptors::{
 };
 use aws_smithy_runtime_api::config_bag::ConfigBag;
 use http::{HeaderName, HeaderValue};
+use std::fmt::Debug;
 use uuid::Uuid;
 
+#[cfg(feature = "test-util")]
+use std::sync::{Arc, Mutex};
+
 #[allow(clippy::declare_interior_mutable_const)] // we will never mutate this
 const AMZ_SDK_INVOCATION_ID: HeaderName = HeaderName::from_static("amz-sdk-invocation-id");
 
-/// This interceptor generates a UUID and attaches it to all request attempts made as part of this operation.
-#[non_exhaustive]
-#[derive(Debug)]
-pub struct InvocationIdInterceptor {
-    id: InvocationId,
+/// A generator for returning new invocation IDs on demand.
+pub trait InvocationIdGenerator: Debug + Send + Sync {
+    /// Call this function to receive a new [`InvocationId`] or an error explaining why one couldn't
+    /// be provided.
+    fn generate(&self) -> Result<InvocationId, BoxError>;
```

Comment:
Should we make this `Result<Option<InvocationId, _>>` so that tests can easily disable the invocation ID entirely? The alternative is to update all the old tests to have the invocation ID in the test connections (and tediously update the signatures on those too).